python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -622,146 +622,121 @@ def _generateGrid(self, exposure, forceEvenSized=False, **kwargs):
-            # Readjust gridStepY so that it fits perfectly in the image.
-            gridStepY = float(bbox.getHeight() - gridSizeY) / float(nGridY)
-            if gridStepY < 1:
-                raise ValueError('Y grid spacing is too small (or negative): %f' % gridStepY)
-
-        if adjustGridOption == 'size':
-            gridSizeX = gridStepX
-            gridSizeY = gridStepY
-
-        print('Grid parameters:', gridSizeX, gridSizeY, gridStepX, gridStepY, borderSizeX, borderSizeY)
+                cellCentroids = [(x, y) for x in np.arange(cellSizeX//2, bbox.getWidth() + cellSizeX//2,
+                                                           gridStepX)
+                                        for y in np.arange(cellSizeY//2, bbox.getHeight() + cellSizeY//2,
+                                                           gridStepY)]
+
+            cellCentroidsX, cellCentroidsY = zip(*cellCentroids)
 
         # first "main" box at 0,0
-        bbox0 = afwGeom.Box2I(afwGeom.Point2I(bbox.getBegin()), afwGeom.Extent2I(gridSizeX, gridSizeY))
+        bbox0 = afwGeom.Box2I(afwGeom.Point2I(bbox.getBegin()), afwGeom.Extent2I(cellSizeX, cellSizeY))
         # first expanded box
         bbox1 = afwGeom.Box2I(bbox0)
         bbox1.grow(afwGeom.Extent2I(borderSizeX, borderSizeY))
 
         self.boxes0 = []  # "main" boxes; store in task so can be extracted if needed
         self.boxes1 = []  # "expanded" boxes
 
-        # use given centroids as centers for bounding boxes
-        if self.config.gridCentroidsX is not None and len(self.config.gridCentroidsX) > 0:
-            for i, centroidX in enumerate(self.config.gridCentroidsX):
-                centroidY = self.config.gridCentroidsY[i]
+        def _makeBoxEvenSized(bb):
+            """Force a bounding-box to have dimensions that are modulo 2."""
+
+            if bb.getWidth() % 2 == 1:  # grow to the right
+                bb.include(afwGeom.Point2I(bb.getMaxX()+1, bb.getMaxY()))  # Expand by 1 pixel!
+                bb.clip(bbox)
+                if bb.getWidth() % 2 == 1:  # clipped at right -- so grow to the left
+                    bb.include(afwGeom.Point2I(bb.getMinX()-1, bb.getMaxY()))
+                    bb.clip(bbox)
+            if bb.getHeight() % 2 == 1:  # grow upwards
+                bb.include(afwGeom.Point2I(bb.getMaxX(), bb.getMaxY()+1))  # Expand by 1 pixel!
+                bb.clip(bbox)
+                if bb.getHeight() % 2 == 1:  # clipped upwards -- so grow down
+                    bb.include(afwGeom.Point2I(bb.getMaxX(), bb.getMinY()-1))
+                    bb.clip(bbox)
+
+            return bb
+
+        # Use given or grid-parameterized centroids as centers for bounding boxes
+        if cellCentroidsX is not None and len(cellCentroidsX) > 0:
+            for i, centroidX in enumerate(cellCentroidsX):
+                centroidY = cellCentroidsY[i]
                 centroid = afwGeom.Point2D(centroidX, centroidY)
```

Comment:
Plus, the grid may not actually be a grid (if the config-ed `cellCentroidsX` and `cellCentroidsY` are not None). So I'm leaving as-is but refactoring a bit more to reduce code duplication.